components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/request/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,411 @@
+        }
+    }
+
+    /**
+     * Returns array of UserChallengeAnswer from the servlet request parameters from the front-end
+     *
+     * @param servletRequest HTTP Servlet Request.
+     * @return Array of UserChallengeAnswer.
+     */
+    private UserChallengeAnswer[] retrieveChallengeQuestionAnswers(HttpServletRequest servletRequest,
+                                                                   List<ChallengeQuestion> list) {
+        Map<String, String> questionsMap = new HashMap<>();
+        Map<String, String> answersMap = new HashMap<>();
+        List<UserChallengeAnswer> questionsAndAnswers = new ArrayList<>();
+        List<String> paramNamesList = new ArrayList<>();
+
+        Enumeration<String> paramNames = servletRequest.getParameterNames();
+        while (paramNames.hasMoreElements()) {
+            paramNamesList.add(paramNames.nextElement());
+        }
+
+        for (String requestParam : paramNamesList) {
+            if (requestParam.contains(SELECTED_CHALLENGE_QUESTION_PREFIX)) {
+                String question = servletRequest.getParameter(requestParam);
+                String questionSetID = requestParam.replace(SELECTED_CHALLENGE_QUESTION_PREFIX, "");
+                questionsMap.put(questionSetID, question);
+
+            } else if (requestParam.contains(CHALLENGE_QUESTION_ANSWER_PREFIX)) {
+                String answer = servletRequest.getParameter(requestParam);
+                String answerSetID = requestParam.replace(CHALLENGE_QUESTION_ANSWER_PREFIX, "");
+                answersMap.put(answerSetID, answer);
+            }
+        }
+        for (String questionKey : questionsMap.keySet()) {
+            String challengeQuestion = questionsMap.get(questionKey);
+            for (ChallengeQuestion question : list) {
+                if (question.getQuestion().equals(challengeQuestion)) {
+                    UserChallengeAnswer questionAndAnswer = new UserChallengeAnswer();
+                    questionAndAnswer.setQuestion(question);
+                    questionAndAnswer.setAnswer(answersMap.get(questionKey));
+                    questionsAndAnswers.add(questionAndAnswer);
+                }
+            }
+        }
+        return questionsAndAnswers.toArray(new UserChallengeAnswer[questionsAndAnswers.size()]);
+    }
+
+    /**
+     * Returns URL encoded String with challenge questions of the given user
+     *
+     * @param user Authenticated User.
+     * @return UTF-8 encoded URL with challenge questions.
+     */
+    private String getUrlEncodedChallengeQuestionsString(AuthenticatedUser user) throws UnsupportedEncodingException {
+        StringBuilder challengeQuestionData = new StringBuilder();
+        List<ChallengeQuestion> challengeQuestionList = getChallengeQuestions(user);
+
+        if (challengeQuestionList == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Challenge Questions for found for tenant domain of the user: " + user.getUserName());
```

Comment:
shall we log the tenant domain as well?
Also typo.. change as "Challenge questions not found for.. "

Reply:
Please reword as.. "Challenge questions not found for the user: user.getUsername() in tenant domain: user.getTenantDomain"